src/components/HistoryTable/historyTableContextReducer.ts

```diff
@@ -0,0 +1,122 @@
+import { processCommits, CommitRowType, mainlineCommits } from "./utils";
+
+type Action =
+  | { type: "ingestNewCommits"; commits: mainlineCommits }
+  | { type: "addColumns"; columns: string[] }
+  | { type: "nextPageColumns" }
+  | { type: "prevPageColumns" }
+  | { type: "setColumnLimit"; limit: number };
+
+type cacheShape = {
+  [order: number]:
+    | mainlineCommits["versions"][0]["version"]
+    | mainlineCommits["versions"][0]["rolledUpVersions"][0];
+};
+interface HistoryTableState {
+  loadedCommits: mainlineCommits["versions"];
+  processedCommits: CommitRowType[];
+  processedCommitCount: number;
+  commitCache: cacheShape;
+  cacheSize: number;
+  visibleColumns: string[];
+  currentPage: number;
+  columns: string[];
+  columnLimit: number;
+}
+
+export const reducer = (state: HistoryTableState, action: Action) => {
+  switch (action.type) {
+    case "ingestNewCommits": {
+      // We cache the commits and use this to determine if a new commit was added in this action
+      //   This also performantly handles deduplication of commits at the expense of memory
+      const updatedObjectCache = objectifyCommits(
+        state.commitCache,
+        action.commits.versions
+      );
+      const updatedCache = Object.values(updatedObjectCache);
+      if (updatedCache.length > state.cacheSize) {
+        const processedCommits = processCommits(
+          action.commits.versions,
+          state.processedCommits
+        );
+        return {
+          ...state,
+          commitCache: updatedObjectCache,
+          cacheSize: updatedCache.length,
+          processedCommits,
+          processedCommitCount: processedCommits.length,
+        };
+      }
+      return {
```

Comment:
we can improve performance by not creating a new reference for state when there are no changes to it. i wonder if this can fix the useEffect involving fetchNewCommit [here](https://github.com/evergreen-ci/spruce/pull/869/files#diff-9b06585f88cd125563a91d3134e8296f02b8e553b1873edf086247fcbbcddb41R23-R28) so we can pass in all deps